src/core/evaluator.js

```diff
@@ -2304,29 +2347,46 @@ class PartialEvaluator {
             textState.leading = args[0];
             break;
           case OPS.moveText:
-            // Optimization to treat same line movement as advance
-            var isSameTextLine = !textState.font
-              ? false
-              : (textState.font.vertical ? args[0] : args[1]) === 0;
-            advance = args[0] - args[1];
             if (
               combineTextItems &&
-              isSameTextLine &&
               textContentItem.initialized &&
-              advance > 0 &&
-              advance <= textContentItem.fakeMultiSpaceMax
+              textState.font
             ) {
-              textState.translateTextLineMatrix(args[0], args[1]);
-              textContentItem.width +=
-                args[0] - textContentItem.lastAdvanceWidth;
-              textContentItem.height +=
-                args[1] - textContentItem.lastAdvanceHeight;
-              diff =
-                args[0] -
-                textContentItem.lastAdvanceWidth -
-                (args[1] - textContentItem.lastAdvanceHeight);
-              addFakeSpaces(diff, textContentItem.str);
-              break;
+              if (textState.font.vertical) {
+                advance = args[1] - textContentItem.lastAdvanceHeight;
+                if (args[0] === 0 /* same column */) {
+                  if (-0.5 * textContentItem.lastCharSize <= advance) {
+                    const isSplited = addFakeSpaces(advance);
+                    textState.translateTextLineMatrix(args[0], args[1]);
+                    if (isSplited) {
+                      textState.textMatrix = textState.textLineMatrix.slice();
+                    } else {
+                      textContentItem.height += advance;
+                    }
+                    break;
+                  }
+                } else if (advance < 0) {
+                  textContentItem.hasEOL = true;
+                }
+              } else {
+                advance = args[0] - textContentItem.lastAdvanceWidth;
+                // Because of kerning a glyph can be moved back but it still
+                // follows the previous one.
+                if (args[1] === 0 /* same line */) {
+                  if (-0.5 * textContentItem.lastCharSize <= advance) {
+                    const isSplited = addFakeSpaces(advance);
+                    textState.translateTextLineMatrix(args[0], args[1]);
+                    if (isSplited) {
+                      textState.textMatrix = textState.textLineMatrix.slice();
+                    } else {
+                      textContentItem.width += advance;
+                    }
+                    break;
+                  }
+                } else if (advance < -2 * textContentItem.lastCharSize) {
```

Comment:
Another magical value (i.e. `-2`), please at least add a comment explaining what it is/does.